### DISCUSSION TITLE
Input validation / XML Schema

### DISCUSSION THREAD

**Discussion by cssherman** [GEOS-DEV/GEOS#1516] [Usability]
Now that we are starting to collect additional users, it may be useful to talk about validating input xml files.  So far, we have been validating xml files that are within the repository when the code is built.  This is great for keeping files up-to-date, but would be onerous for typical users.

There are a variety of tools that can be used in text editors to validate files on the fly (see https://geosx-geosx.readthedocs-hosted.com/en/latest/coreComponents/fileIO/doc/InputValidation.html#input-file-validation).  To work, these require a couple of additional attributes in the problem node:

```xml
<Problem xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
xsi:schemaLocation="/g/g17/sherman/GEOS/geosx/GEOSX/src/coreComponents/schema/schema.xsd">
...
</Problem>
```

Some of the issues that we face with this approach include:

1) The schemaLocation attribute is currently pointing to a local path.  For this to be really useful, we would need to host the current schema somewhere online, and configure the current xml files to look for it.
2) While the schema is relatively stable these days, we may run into issues where users try to validate files against older versions.
3) In certain environments, the user would not have access to the internet in order to grab the schema.  This would require them to point their editor to a local copy of the schema or to use a command line tool such as xmllint or preprocess_xml.

**Comment by TotoGaz** [comment]
Isn't `geosx` able to generate its own `schema.xsd` file? (`geosx -s schema.xsd`)
Then it would just be a matter of `xsd` validation and proper documentation. Nothing we have to implement ourselves.

**Reply by cssherman** [reply]
It's certainly a good option to have as a backup.  Looking around, there are some preexisting tools that can fetch the current raw file from a git repository (which takes up the role of the old raw.github.com tool).  For example, this could work: https://www.jsdelivr.com/?docs=gh .  (Obviously, we'd want to vet any service we use)

**Reply by cssherman** [reply]
On second thought, after playing around with thing a bit, I think that generating a schema locally would be preferred.   I can write some additional instructions for how to do this in the documentation.

**Reply by TotoGaz** [reply]
That would be great, thx!

Now that our `travis` is properly ensuring that `geosx` will be able to generate it's own scheme, this will be a simple and efficient way to ensure a consistent check!

> I can write some additional instructions for how to do this in the documentation.

Or in our FAQ! 😉 

Also:
```
pugixml enables very fast, [...] also the parser is a non-validating one, so if you need DTD/Schema validation, the library is not for you.
```
That's bad luck, some `geosx --check input.xml` would have been the most simple for the users!
